### PR TITLE
Automate Javadoc shortlinks generation

### LINKED ISSUE #4023: Automate generation of https://guava.dev/Foo Javadoc shortlinks
I think I generated them once by hand. Generating them automatically during builds shouldn't be too hard.

(I noticed because I just wanted to reference https://guava.dev/ClosingFuture and was reminded that that won't exist until we add it ourselves.)

I may have some other notes about this somewhere. Certainly I have at least some on internal bug 157460102, which is about adding more shortlink conveniences and Javadoc munging.

### PR DESCRIPTION
**Pull request by benrkia**
Fixes #4023

### CODE REVIEW

File: updaterelease.sh [hunk 1]
```diff
@@ -358,6 +358,31 @@ update_config_yml() {
   fi
 }
 
+# Generates Javadoc shortlinks for the current snapshot release.
+generate_snapshot_javadoc_shortlinks() {
+  # Creates 2 sub-folders for each non-nested class inside javadocshortcuts.
+  # This ensures that each class documentation is accessible using  https://guava.dev/ClassName
+  # or https://guava.dev/classname.
+  if [[ "$RELEASE" == "snapshot" ]]; then
```
Comment (netdpb) [inline updaterelease.sh:366]: Might be easier to read if you invert this.
```suggestion
  if [[ "$RELEASE" != "snapshot" ]]; then
    return
  fi
```

File: updaterelease.sh [hunk 2]
```diff
@@ -363,24 +363,25 @@ generate_snapshot_javadoc_shortlinks() {
...
+    mkdir -p javadocshortcuts/{$SHORT,$SHORT_LOWER} && (
+      echo ---
+      echo "title: $SHORT"
+      echo "permalink: /$SHORT/"
+      echo "redirect_to: https://guava.dev/$F"
+      echo "---"
+    ) | tee javadocshortcuts/{$SHORT,$SHORT_LOWER}/index.md >/dev/null
+
+    # Sets the permalink value to lowercase for the 2nd sub-folder.
+    perl -pi -e '$_ = lc $_ if /^permalink/' javadocshortcuts/$SHORT_LOWER/index.md
```
Comment (netdpb) [inline updaterelease.sh:383]:
```suggestion
    perl -pi -e 's/^permalink.*/\L$&/' javadocshortcuts/$SHORT_LOWER/index.md
```

File: updaterelease.sh [hunk 3]
```diff
@@ -363,24 +363,25 @@ generate_snapshot_javadoc_shortlinks() {
...
+    SHORT=$(basename $F .html)
+
+    # Lowercases the 2nd sub-folder's name
+    SHORT_LOWER=$(echo $SHORT | tr A-Z a-z)
+    mkdir -p javadocshortcuts/{$SHORT,$SHORT_LOWER} && (
+      echo ---
+      echo "title: $SHORT"
+      echo "permalink: /$SHORT/"
+      echo "redirect_to: https://guava.dev/$F"
+      echo "---"
```
Comment (netdpb) [inline updaterelease.sh:379]:
```suggestion
      echo ---
```

File: updaterelease.sh [hunk 4]
```diff
@@ -358,6 +358,32 @@ update_config_yml() {
...
 
+# Generates Javadoc shortlinks for the current snapshot release.
+generate_snapshot_javadoc_shortlinks() {
+  # Creates 2 sub-folders for each non-nested class inside javadocshortcuts.
+  # This ensures that each class documentation is accessible using  https://guava.dev/ClassName
+  # or https://guava.dev/classname.
+  if [[ "$RELEASE" != "snapshot" ]]; then
+    return
+  fi
+  for F in $(find releases/snapshot/api/docs/com/google/common -name '[A-Z]*.html' -not -path '*/class-use/*' -not -name '*.*.*'); do
```
Comment (cpovirk) [inline updaterelease.sh:369]:
```suggestion
  for F in $(find releases/snapshot-jre/api/docs/com/google/common -name '[A-Z]*.html' -not -path '*/class-use/*' -not -name '*.*.*'); do
```

Comment (cpovirk) [reply updaterelease.sh:369]: (What you have should work, but we generally link to `snapshot-jre` nowadays for consistency.)

**Review by cpovirk — APPROVED** [review]
Thanks!

**Comment by cpovirk** [comment]
@netdpb , do you want me to create an internal copy of this and mail it to you for review?

**Comment by netdpb** [comment]
> @netdpb , do you want me to create an internal copy of this and mail it to you for review?

Sure, that would be great. Thanks.

**Comment by cpovirk** [comment]
Ah, the joke is on me: This file doesn't exist internally :) (It's part of the gh-pages branch, which could be synced (at least in part -- we wouldn't do it for the generated Javadoc) but isn't.)

So we can probably use that "Rebase and merge" button for a change. Sound good?

**Comment by cpovirk** [comment]
It worked, generating a new redirect: https://guava.dev/ClosingFuture

Thanks!

**Comment by benrkia** [comment]
@cpovirk Thank you for giving me the chance to contribute 🙏

**Pull request merged**